# Eigen vector and Eigen values

In [25]:
import numpy as np

---


## Eigen 

$$ 

A \cdot v = \lambda \cdot v

$$

where $A$ is a square matrix, $v$ is a vector and $\lambda$ is a scalar.

if we have a set of vectors $v_1, v_2, v_3, ... v_n$ and a set of scalars $\lambda_1, \lambda_2, \lambda_3, ... \lambda_n$ such that

$$

A \cdot v_1 = \lambda_1 \cdot v_1 \\

A \cdot v_2 = \lambda_2 \cdot v_2  \\

A \cdot v_3 = \lambda_3 \cdot v_3  \\

$$

then $v_1, v_2, v_3, ... v_n$ are called the eigen vectors of matrix $A$ and $\lambda_1, \lambda_2, \lambda_3, ... \lambda_n$ are called the eigen values of matrix $A$.

$V = [v_1, v_2, v_3, ... v_n]$ and $\Lambda = diag([\lambda_1, \lambda_2, \lambda_3, ... \lambda_n])$

$$

A \cdot V = V \cdot \Lambda

$$

if $V is invertible which means Rank(V) = n and $V^{-1}$ exists then

$$

A = V \cdot \Lambda \cdot V^{-1} \\

\Lambda = V^{-1} \cdot A \cdot V
$$

---

> But you also need to be comfortable with determinants , linear systems of equations, and change of basis. Confusion about eigen-stuffs usually has more to do with a shaky foundation in one of these topics than it does with the eigenvectors and eigenvalues themselves.


Prerequisites:

- [Determinants]
- [Linear Systems of Equations]
- [Change of Basis]

## An example

To start, consider some linear tranformation in two dimensions, like the one shown here.

It moves the basis vector $\hat{i}$ to the coordinates $\left[ \begin{array}{c} 3 \\ 0 \end{array} \right]$ and the basis vector $\hat{j}$ to the coordinates $\left[ \begin{array}{c} 0 \\ 2 \end{array} \right]$, so it's represented with a matrix $\left[ \begin{array}{cc} 3 & 0 \\ 0 & 2 \end{array} \right]$.

Now, focus on what it does to any individual vector, and think about the span of that vector, which is the line passing through the origin and its tip.

Most vectors will get knocked off that line during the transformation. It would seem pretty coincidental if the place where a vector lands happens to be somewhere on that line. But some special vectors do remain on their own span, meaning the effect the matrix has on such a vector is to just stretch or squish it, like a scalar.

For this specific example, the basis vector $\hat{i}$ is one such vector that with this special property. The span of $\hat{i}$ is the x-axis, and from the first column of the matrix we can see that $\hat{i}$ move over to 3 times itself, still on the x-axis. Waht's more, because of the way linear transformations work, any other vector on the x-axis will also just get stretched or squished by 3, and hence remains on its own span.

A slightly sneakier vector that remains on its own span during this transformation is $\left [ \begin{array}{c} 1 \\ 1 \end{array} \right]$. It ends up getting stretched out by a factor of 2. Again, linearity will imply that any other vector on the diagonal line spanned by this one will just get stretched out by a factor of 2. 

In [29]:
A = np.matrix([[3, 1], [0,2]])

print(A)

# v1 is a column vector
v1 = np.matrix([[1], [0]])

print(v1)

v11 = A.dot(v1)

print(v11)

[[3 1]
 [0 2]]
[[1]
 [0]]
[[3]
 [0]]


Most vectors will get knocked off that line during the transformation. It would seem pretty coincidental if the place where a vector lands happens to be somewhere on that line. But some special vectors do remain on their own span, meaning the effect the matrix has on such a vector is to just stretch or squish it, like a scalar.

In [34]:
A = np.matrix([[3, 1], [0, 2]])

print(A)

# v1 is a column vector
v2 = np.matrix([[1], [-1]])

print(v2)

v22 = A.dot(v2)

print(v22)


[[3 1]
 [0 2]]
[[ 1]
 [-1]]
[[ 2]
 [-2]]


## 3d rotational axis

For a glimpse of why this might be a useful thing to think about, consider some three-dimensional rotation. If you can find an eigenvector for this rotation, a vector which remains on its own span, what you have found is the axis of rotation

![3d-rotational-axis](./AxisOfRotationEigenvector.light.svg)

Squishification $\det(A - \lambda I) = 0$

The matrix $A - \lambda I $ squishes space onto a line. That means there's a nonzero vector $v$ such that 

$$

(A - \lambda I) \cdot v = 0

## Rotation

Now, a 2d transformation doesn't have to have eigenvectors. For example, consider a rotation by 
90 degrees. This doesn't have any eigenvectors, since each vector is rotated off of its own span.

In [35]:
R = np.matrix([[0, -1], [1, 0]])

print(R)

# get eigenvalue 

eigenvalues, eigenvectors = np.linalg.eig(R)

print(eigenvalues)

print(eigenvectors)

[[ 0 -1]
 [ 1  0]]
[0.+1.j 0.-1.j]
[[0.70710678+0.j         0.70710678-0.j        ]
 [0.        -0.70710678j 0.        +0.70710678j]]


The only roots of this polynomial are the imaginary numbers $i$ and $-i$. That means there are no real number indicates that there are no eigenvecotrs for this transformation.

## Shear 

Another interesting example is a shear. This fixes $\hat{i}$ in place, and moves $\hat{j}$ one over, so it's matrix is $\left[ \begin{array}{cc} 1 & 1 \\ 0 & 1 \end{array} \right]$.

In [ ]:
S = np.matrix([[1, 1], [0, 1]])

print(S)

# get eigenvalue

eigenvalues, eigenvectors = np.linalg.eig(S)

print(eigenvalues) # acutually the eigenvalues are 1

print(eigenvectors) # actually the eigenvectors are [1, 0], others are linearly dependent

e1 = np.matrix([[1], [0]])

print(e1)

res1 = S.dot(e1)

print(res1)

[[1 1]
 [0 1]]
[1. 1.]
[[ 1.00000000e+00 -1.00000000e+00]
 [ 0.00000000e+00  2.22044605e-16]]
[[1]
 [0]]
[[1]
 [0]]


## Scaling

It's also possible to have just one eigenvalue with more than just a line full of eigenvectors. A simple example is the matrix which scales everything by 
2, but every vector in the plane gets to be an eigenvector with that eigenvalue.

In [40]:
M = np.matrix([[2, 0], [0, 2]])

print(M)

# get eigenvalue

eigenvalues, eigenvectors = np.linalg.eig(M)

print(eigenvalues)

print(eigenvectors)

[[2 0]
 [0 2]]
[2. 2.]
[[1. 0.]
 [0. 1.]]


## Fibonacci Matrix

---


### **1. 起点**
我们有初始向量：
$
\begin{bmatrix} a \\ b \end{bmatrix}
$
接着通过一系列的递推规则生成新向量。


### **2. 递推规则**
每次变换遵循如下矩阵递推关系：
$
\begin{bmatrix} a \\ b \end{bmatrix} \to \begin{bmatrix} b \\ a + b \end{bmatrix}
$

#### **矩阵形式的表示**
将上述递推写成矩阵形式：
$
\begin{bmatrix} a_{k+1} \\ b_{k+1} \end{bmatrix} =
\begin{bmatrix} 0 & 1 \\ 1 & 1 \end{bmatrix} 
\begin{bmatrix} a_k \\ b_k \end{bmatrix}
$
其中：
- 初始向量为 $\begin{bmatrix} a \\ b \end{bmatrix}$。
- 矩阵 $ \begin{bmatrix} 0 & 1 \\ 1 & 1 \end{bmatrix} $ 是变换矩阵。


### **3. 递推过程**
从初始值 $ a = 1, b = 1 $ 开始：

1. 初始向量：
   $
   \begin{bmatrix} 1 \\ 1 \end{bmatrix}
   $

2. 第一次变换：
   $
   \begin{bmatrix} 0 & 1 \\ 1 & 1 \end{bmatrix} 
   \begin{bmatrix} 1 \\ 1 \end{bmatrix} =
   \begin{bmatrix} 1 \\ 2 \end{bmatrix}
   $

3. 第二次变换：
   $
   \begin{bmatrix} 0 & 1 \\ 1 & 1 \end{bmatrix} 
   \begin{bmatrix} 1 \\ 2 \end{bmatrix} =
   \begin{bmatrix} 2 \\ 3 \end{bmatrix}
   $

4. 第三次变换：
   $
   \begin{bmatrix} 0 & 1 \\ 1 & 1 \end{bmatrix} 
   \begin{bmatrix} 2 \\ 3 \end{bmatrix} =
   \begin{bmatrix} 3 \\ 5 \end{bmatrix}
   $

5. 第四次变换：
   $
   \begin{bmatrix} 0 & 1 \\ 1 & 1 \end{bmatrix} 
   \begin{bmatrix} 3 \\ 5 \end{bmatrix} =
   \begin{bmatrix} 5 \\ 8 \end{bmatrix}
   $


### **4. 递推结果**
这一过程生成了斐波那契数列的前几项：
$
\begin{bmatrix} 1 \\ 1 \end{bmatrix} \to 
\begin{bmatrix} 1 \\ 2 \end{bmatrix} \to 
\begin{bmatrix} 2 \\ 3 \end{bmatrix} \to 
\begin{bmatrix} 3 \\ 5 \end{bmatrix} \to 
\begin{bmatrix} 5 \\ 8 \end{bmatrix} \to 
\begin{bmatrix} 8 \\ 13 \end{bmatrix} \to \cdots
$



### **5. 一般公式**
根据矩阵递推的特性，可以写出递推关系的一般公式：
$
\begin{bmatrix} a_k \\ b_k \end{bmatrix} = 
\begin{bmatrix} 0 & 1 \\ 1 & 1 \end{bmatrix}^k 
\begin{bmatrix} a_0 \\ b_0 \end{bmatrix}
$
- 初始条件 $ \begin{bmatrix} a_0 \\ b_0 \end{bmatrix} $ 可以选择任意值。

对于斐波那契数列，初始值为：
$
\begin{bmatrix} a_0 \\ b_0 \end{bmatrix} = \begin{bmatrix} 1 \\ 1 \end{bmatrix}
$



In [3]:
import numpy as np

# 计算A的特征值和特征向量
A = np.matrix([[0, 1], [1, 1]])


eigenvalues, eigenvectors = np.linalg.eig(A)

print(eigenvalues)

print(eigenvectors)

[-0.61803399  1.61803399]
[[-0.85065081 -0.52573111]
 [ 0.52573111 -0.85065081]]


矩阵的特征向量为: $ \phi = 1.618033988749895 $ 和 $ \phi = -0.6180339887498949 $

### 黄金分割比（Golden Ratio, $ \phi $）

黄金分割比 $ \phi $ 是数学中的一个著名常数，约等于 $ 1.618 \dots $。它经常出现在几何学、艺术、自然和数列中，尤其是与斐波那契数列密切相关。



### **1. 黄金分割比的定义**

黄金分割比满足以下方程：
$
\phi = \frac{1 + \sqrt{5}}{2} \approx 1.618...
$

#### 推导：
假设一个线段被分割为两部分 $ a $ 和 $ b $（其中 $ a > b > 0 $），使得：
$
\frac{a}{b} = \frac{a+b}{a}
$

解这个比例方程可以得到：
$
\phi = \frac{1 + \sqrt{5}}{2} \quad \text{（正解）}
$
负解 $ \frac{1 - \sqrt{5}}{2} \approx -0.618 $ 也是该方程的解，但通常不用于表示黄金比。



### **2. 反黄金分割比（$ -\frac{1}{\phi} $）**

#### 定义：
黄金分割比的一个重要性质是其倒数：
$
\frac{1}{\phi} = \phi - 1 \approx 0.618...
$
反黄金分割比则是：
$
-\frac{1}{\phi} = -\phi + 1 \approx -0.618...
$

#### 解释：
- $ \phi $ 和 $ -\frac{1}{\phi} $ 是相关联的值，分别对应黄金分割中的“正比例”和“负比例”。
- 在斐波那契数列的生成公式中，这两个值也起到了关键作用（见下文）。



### **3. 黄金分割与斐波那契数列**

黄金分割比出现在斐波那契数列中。斐波那契数列是一个递归数列：
$
F_n = F_{n-1} + F_{n-2}, \quad F_0 = 0, \, F_1 = 1
$

对于较大的 $ n $，斐波那契数列的第 $ n $ 项 $ F_n $ 与第 $ n-1 $ 项的比值会趋近于黄金分割比：

$$
\frac{F_n}{F_{n-1}} \to \phi \quad \text{as } n \to \infty
$$




### **4. 二次方程的关系**

黄金分割比和反黄金分割比是以下二次方程的解：
$
x^2 - x - 1 = 0
$
解这个方程得到：
$
x = \frac{1 + \sqrt{5}}{2} \quad \text{和} \quad x = \frac{1 - \sqrt{5}}{2}
$
它们分别是：
- 黄金分割比：$ \phi = \frac{1 + \sqrt{5}}{2} \approx 1.618... $
- 反黄金分割比：$ -\frac{1}{\phi} = \frac{1 - \sqrt{5}}{2} \approx -0.618... $


### **5. 总结**

- 黄金分割比 $ \phi \approx 1.618 $ 是数学中极为重要的一个常数，表示一种特殊的比例。
- 反黄金分割比 $ -\frac{1}{\phi} \approx -0.618 $ 是其负倒数，具有对称性质。
- 它们在斐波那契数列、几何学和二次方程中扮演重要角色，是数学中优雅而深刻的一个现象。


### **规范的公式表示**
规范化的公式版本：

$
A_n = S D_n S^{-1}
$

其中：
- $ S $：特征向量矩阵。
- $ D_n $：特征值矩阵，形式为对角矩阵。
- $ S^{-1} $：特征向量矩阵的逆。

---

### **公式展开**
1. **初始定义**：
   $
   S = \frac{1}{5} \begin{bmatrix} -\lambda_2 & 1 \\ -\lambda_1 & 1 \end{bmatrix}, \quad
   D_n = \begin{bmatrix} \lambda_1^n & 0 \\ 0 & \lambda_2^n \end{bmatrix}, \quad
   S^{-1} = \begin{bmatrix} 1 & -1 \\ \lambda_1 & -\lambda_2 \end{bmatrix}
   $

2. **递推展开**：
   $
   A_n = S D_n S^{-1}
   $
   展开为：
   $
   A_n = \frac{1}{5} 
   \begin{bmatrix} 
   -\lambda_2 & 1 \\ 
   -\lambda_1 & 1 
   \end{bmatrix}
   \begin{bmatrix} 
   \lambda_1^n & 0 \\ 
   0 & \lambda_2^n 
   \end{bmatrix}
   \begin{bmatrix} 
   1 & -1 \\ 
   \lambda_1 & -\lambda_2 
   \end{bmatrix}
   $

3. **化简计算**：
   逐步相乘：
   - 第一步：
     $$
     \begin{bmatrix} 
     -\lambda_2 & 1 \\ 
     -\lambda_1 & 1 
     \end{bmatrix}
     \begin{bmatrix} 
     \lambda_1^n & 0 \\ 
     0 & \lambda_2^n 
     \end{bmatrix}
     =
     \begin{bmatrix} 
     -\lambda_2 \lambda_1^n & \lambda_2^n \\ 
     -\lambda_1 \lambda_1^n & \lambda_2^n 
     \end{bmatrix}
     $$

   - 第二步：
     $$
     \begin{bmatrix} 
     -\lambda_2 \lambda_1^n & \lambda_2^n \\ 
     -\lambda_1 \lambda_1^n & \lambda_2^n 
     \end{bmatrix}
     \begin{bmatrix} 
     1 & -1 \\ 
     \lambda_1 & -\lambda_2 
     \end{bmatrix}
     =
     \begin{bmatrix} 
     \lambda_1^{n-1} - \lambda_2^{n-1} & \lambda_1^n - \lambda_2^n \\ 
     \lambda_1^n - \lambda_2^n & \lambda_1^{n+1} - \lambda_2^{n+1}
     \end{bmatrix}
     $$

---

### **最终结果**
经过化简，递推公式可以写为：
$
A_n = \frac{1}{5} 
\begin{bmatrix} 
\lambda_1^{n-1} - \lambda_2^{n-1} & \lambda_1^n - \lambda_2^n \\ 
\lambda_1^n - \lambda_2^n & \lambda_1^{n+1} - \lambda_2^{n+1}
\end{bmatrix}
$

---

### **解释**
1. **特征值分解的意义**：
   矩阵 $ A_n $ 是通过特征值矩阵 $ D_n $ 以及特征向量矩阵 $ S $ 的分解构造的。特征值矩阵 $ D_n $ 的对角项为 $ \lambda_1^n $ 和 $ \lambda_2^n $，代表指数增长或衰减的趋势。

2. **与斐波那契数列的关系**：
   - 如果特征值 $ \lambda_1 $ 和 $ \lambda_2 $ 是黄金分割比和其逆数，矩阵 $ A_n $ 中的项就会与斐波那契数列的递推关系一致。
   - 即：矩阵中元素会递推生成斐波那契数列。

3. **用途**：
   - 这种分解方法在递推数列、差分方程、动力系统以及数值计算中非常重要。

---

可以得出Fibonacci数列的通项公式为：

$$
F_n = \frac{\lambda_1^n - \lambda_2^n}{\sqrt{5}}, \quad \text{其中} \quad \lambda_1 = \frac{1 + \sqrt{5}}{2}, \, \lambda_2 = \frac{1 - \sqrt{5}}{2}
$$